### PR TITLE
[EVM-Equivalence-YUL] Restrict functions to only be called from interpreter

### PR DESCRIPTION
**Pull request by gianbelinche**
# What ❔
This PR adds system call modifiers to some system contracts functions to prevent security problems.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

**Comment by 0xVolosnikov** [comment]
I think it's better to remove `createEVMInternal`, remove a special right for evm contracts to call `increaseDeploymentNonce` in nonce holder and just use encapsulated functionality from `ContractDeployer`

File: system-contracts/contracts/ContractDeployer.sol [hunk 1]
```diff
@@ -37,7 +37,16 @@ contract ContractDeployer is IContractDeployer, ISystemContract {
 
     uint256 public constructorReturnGas;
 
-    function setDeployedCode(uint256 constructorGasLeft, bytes calldata paddedNewDeployedCode) external {
+    modifier onlySystemEvm() {
+        require(ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT.isAccountEVM(msg.sender), "only system evm");
+        require(
+            SystemContractHelper.isSystemCall() || SystemContractHelper.isSystemContract(msg.sender),
```
Comment (0xVolosnikov) [inline system-contracts/contracts/ContractDeployer.sol:43]: Caller can't be a system contract

File: system-contracts/contracts/EvmGasManager.sol [hunk 2]
```diff
@@ -79,6 +82,10 @@ contract EvmGasManager {
 
     modifier onlySystemEvm() {
         require(ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT.isAccountEVM(msg.sender), "only system evm");
+        require(
+            SystemContractHelper.isSystemCall() || SystemContractHelper.isSystemContract(msg.sender),
```
Comment (0xVolosnikov) [inline system-contracts/contracts/EvmGasManager.sol:86]: Caller can't be a system contract

File: system-contracts/contracts/NonceHolder.sol [hunk 3]
```diff
@@ -132,12 +132,13 @@ contract NonceHolder is INonceHolder, ISystemContract {
     /// @notice Increments the deployment nonce for the account and returns the previous one.
     /// @param _address The address of the account which to return the deploy nonce for.
     /// @return prevDeploymentNonce The deployment nonce at the time this function is called.
-    function incrementDeploymentNonce(address _address) external returns (uint256 prevDeploymentNonce) {
+    function incrementDeploymentNonce(address _address) external onlySystemCall returns (uint256 prevDeploymentNonce) {
```
Comment (0xVolosnikov) [inline system-contracts/contracts/NonceHolder.sol:135]: Maybe it is a bad idea. Probably we should change a require below instead

File: system-contracts/contracts/EvmGasManager.sol (hunk 2 above)
Comment (gianbelinche) [reply system-contracts/contracts/EvmGasManager.sol]: Done!

File: system-contracts/contracts/NonceHolder.sol (hunk 3 above)
Comment (gianbelinche) [reply system-contracts/contracts/NonceHolder.sol]: With the change now using createEVM, that modifier is not needed anymore.

**Comment by gianbelinche** [comment]
> I think it's better to remove `createEVMInternal`, remove a special right for evm contracts to call `increaseDeploymentNonce` in nonce holder and just use encapsulated functionality from `ContractDeployer`

Done

**Comment by 0xVolosnikov** [comment]
Do we have gas benchmarks and / or integration tests passing?

**Comment by gianbelinche** [comment]
> Do we have gas benchmarks and / or integration tests passing?

Both compiler tester tests (semanticTests) and integration tests are passing

**Review by 0xVolosnikov — APPROVED** [review]

**Comment by 0xVolosnikov** [comment]
please run prettier for contracts

**Pull request merged**